### ISSUE TITLE
Stop debuging process

### ISSUE THREAD

**Issue by acotty** [eranif/dbgd#2]
I may have missed something or may be looking at this in the wrong way, but here goes:

I can only see in the DAP code a m_client.Reset() function that you can use to stop the debugger (not pause), but it does not generate any event to indicate that the debuggee has stopped or not.

I was hoping the process was like the startup where the following process occurs:
     -> send m_client.Launch()
     <- receive the wxEVT_DAP_LAUNCH_RESPONSE event
Like the following:
     -> send m_client.StopDebugging() or use the existing m_client.Reset()
     <- receive the wxEVT_DAP_TERMINATED_EVENT event or a new event

At this point in time, which may change, that the OnTerminated(...) function would contain the GUI cleanup code for any debugee termination, but at the moment I need to have to duplicate it in the stopdebugging code that is hooked up to the sop gui bitton and also in the OnTerminated() function unless I create yet another function.

So What do you see as the correct way kill the debuggee effectively from a wxWidget stop debugging button?

BTW: Do you want these type of issues/questions in this repo as an issue or should I use the CodeLite forum or if you enable discussions for this github repo then they could be don in teh discussion area or is there some other way?

**Comment by eranif** [comment]
Here is fine. I don't mind discussions here

I usually prefer the `m_client.Reset()` since I don't really want to rely on an external implementation to implement this for me, consider the option that the process might crash or fail to follow the protocol to the letter?

In my implementation within CodeLite, I handle all cases like this:

- [OnDapLaunchResponse][1] with `success == false` I call to an internal method `DoCleanup()`
- [OnDebugStop()][2] - when user clicks the `Stop` button, I call to `DoCleanup()`
- [OnDapExited][3] - same, I call `DoCleanup()`

`DoCleanup()` does the following:
- Removes the debugger marker from the UI
- calls `m_client.Reset()`
- Terminates the dap server (`codelldb` or `lldb-vscode` or whatever)

Some notes and observations from testing it against 2 servers:

1.
You should note that I am testing the demo against two dap-servers: `codelldb` and `lldb-vscode` (my personal favorite)
There are problems, but in general, lldb-vscode (from the LLVM team, the people how wrote clang and clangd) seems the better and more reliable to me

2.
There is an issue with setting breakpoints, you can add function breakpoint (click on the new button I added to the UI and type the function name). However, placing breakpoints with file:line format does not work reliably. It only works when passing relative file path (e.g main.cpp instead of C:\src\main.cpp) - this looks like a bug to me (lldb-vscode can handle full paths, but it requires some strange format like `/Users/eran/src/main.cpp` instead of native format like `C:\Users\eran\src\main.cpp`)

[1]: https://github.com/eranif/codelite/blob/master/DebugAdapterClient/DebugAdapterClient.cpp#L891
[2]: https://github.com/eranif/codelite/blob/master/DebugAdapterClient/DebugAdapterClient.cpp#L462
[3]: https://github.com/eranif/codelite/blob/master/DebugAdapterClient/DebugAdapterClient.cpp#L878

**Comment by acotty** [comment]
Thanks very much for the post. 
I found some issues with the codelldb and now that I have started to use lldb-vscode I have found it to be more reliable. 

I have just started to add breakpoint support and have to modify the code to make the path relative as I am using the GDB/MI code as a starting point.

The lldb-vscode path looks like a CygWin path and I had to do this conversion last year when I updated the Cygwin debugging support to work in C::B. If you look in the https://github.com/acotty/CodeBlocks_Unofficial_Testing/blob/master/src/sdk/cygwin.cpp file and look at the GetCygwinPath() function that I did to make the conversion. Unfortunately I used the Cygwin cygpath.exe so it will not work, but will give you some pointers if you do want to eventually use lldb-vscode.
